activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -421,15 +421,16 @@ def indexes(table_name, name = nil) #:nodoc:
 
       # Returns an array of +Column+ objects for the table specified by +table_name+.
       def columns(table_name) # :nodoc:
-        sql = "SHOW FULL FIELDS FROM #{quote_table_name(table_name)}"
-        execute_and_free(sql, 'SCHEMA') do |result|
-          each_hash(result).map do |field|
-            type_metadata = fetch_type_metadata(field[:Type], field[:Extra])
-            if type_metadata.type == :datetime && field[:Default] == "CURRENT_TIMESTAMP"
-              new_column(field[:Field], nil, type_metadata, field[:Null] == "YES", field[:Default], field[:Collation])
-            else
-              new_column(field[:Field], field[:Default], type_metadata, field[:Null] == "YES", nil, field[:Collation])
-            end
+        table_name = table_name.to_s
+        column_definitions(table_name).map do |field|
+          type_metadata = fetch_type_metadata(field[:Type], field[:Extra])
+          if type_metadata.type == :datetime && field[:Default] == "CURRENT_TIMESTAMP"
+            default, default_function = nil, field[:Default]
+          else
+            default, default_function = field[:Default], nil
+          end
+          new_column(field[:Field], default, type_metadata, field[:Null] == "YES", default_function, field[:Collation]).tap do |column|
+            column.instance_variable_set(:@table_name, table_name)
```

Comment:
can we add a attribute writter to `table_name` or actually pass it in the construtor? Seems we are always setting it after instantiating the object.